### PR TITLE
Use typing.get_type_hints for input/output and resource/config type inference

### LINKED ISSUE #14571: Importing annotations from __future__ breaks Config
### Dagster version

1.3.6

### What's the issue?

We are attempting to create classes which define their own Config structure internally. As such, the model needs to be able to self-reference, which requires that we use `from __future__ import annotations`, but dagster is very much not happy about this!

### What did you expect to happen?

It should be possible to enable forward references (which work fine with Pydantic models) and have them work with `dagster.Config`.

### How to reproduce?

Consider the following:
```
from pydantic import BaseModel
from __future__ import annotations

class Foo(BaseModel):
    num: int
    config: Foo.ModelConfig
    
    class ModelConfig(BaseModel):
        mult: int

    def calculate(self):
        return self.num * self.config.mult
```

This is a valid Pydantic model, where we include the code required to configure the model `Foo` within the definition for Foo. Ideally we'd be able to use the `Foo.BaseModel` as a config directly (maybe a feature request needed!) so external libraries don't need Dagster as a dependency, but we'll adjust accordingly:

```
from dagster import Config, asset
from pydantic import BaseModel
from __future__ import annotations


class Foo(BaseModel):
    num: int
    config: Foo.ModelConfig
    
    class ModelConfig(Config):
        mult: int

    def calculate(self, config):
        return self.num * config.mult
```

Again, all good so far. However when we use this in an asset, we get an error.

```
from dagster import Config, asset
from pydantic import BaseModel
from __future__ import annotations


class Foo(BaseModel):
    num: int
    config: Foo.ModelConfig
    
    class ModelConfig(Config):
        mult: int

    def calculate(self, config):
        return self.num * config.mult
    
@asset
def my_asset(config: Foo.ModelConfig) -> int:
    foo = Foo()
    return foo.calculate(config = config)

___

DagsterInvalidPythonicConfigDefinitionError: 
Error defining Dagster config class.
Unable to resolve config type 'Foo.ModelConfig' to a supported Dagster config type.
```

Ok, let's strip it back to this:

```
from dagster import Config, asset

class ModelConfig(Config):
    mult: int

@asset
def my_asset(config: ModelConfig) -> int:
    foo = Foo()
    return foo.calculate(config = config)
```

This works, but once we enable annotations, we get the same error as before:

```
from dagster import Config, asset
from __future__ import annotations


class ModelConfig(Config):
    mult: int

@asset
def my_asset(config: ModelConfig) -> int:
    foo = Foo()
    return foo.calculate(config = config)

___

DagsterInvalidPythonicConfigDefinitionError: 
Error defining Dagster config class.
Unable to resolve config type 'Foo.ModelConfig' to a supported Dagster config type.

```

### Deployment type

Dagster Cloud

### Deployment details

_No response_

### Additional information

_No response_

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Fixes #14571 

Currently, `Parameter.annotation` is used when inferring the dagster type from an op compute function signature. This causes an error if either: (a) a string is used directly by the user for the parameter annotation; (b) `from __future__ import annotations` is used. This is because both cases cause the value of `Parameter.annotation` to be a string, and our dagster type resolution logic does not handle strings.

This PR replaces direct access to `Parameter.annotation` with `typing.get_type_hints`, which handles evaluation of a string annotation into an actual class which can be handled by our dagster type resolution logic. Note that this only works if the referents of a string annotation is module-level scope-- it will fail for a local scope.

Still this is an upgrade and will fix the problem in most cases. It will:

- Allows users to use string annotations or `from __future__ import annotations` together with dagster type signature inference
- Allows us to use `from __future__ import annotations` in our own code, which is necessary to allow Ruff to auto-sort typing-only imports-- see https://github.com/dagster-io/dagster/pull/14675. Not clear whether this is a good idea.

## How I Tested These Changes

New unit test using string annotations.

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #14707** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/14707" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/14707?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/inference.py [hunk 1]
```diff
@@ -55,11 +63,12 @@ def _infer_output_description_from_docstring(fn: Callable) -> Optional[str]:
...
-    sig = signature(fn)
-
-    annotation = Parameter.empty
-    if not isgeneratorfunction(fn):
-        annotation = sig.return_annotation
+    type_hints = get_type_hints(fn)
+    annotation = (
+        type_hints["return"]
+        if not isgeneratorfunction(fn) and "return" in type_hints
+        else Parameter.empty
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/definitions/inference.py:70]: whats going on here?

File: python_modules/dagster/dagster/_core/decorator_utils.py [hunk 2]
```diff
@@ -36,6 +51,11 @@ def get_function_params(fn: Callable[..., Any]) -> Sequence[Parameter]:
     return list(signature(fn).parameters.values())
 
 
+def get_type_hints(fn: Callable) -> Mapping[str, Any]:
+    target = fn.func if isinstance(fn, functools.partial) else fn
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/decorator_utils.py:55]: is this just to deal with a difference in behavior between `typing_get_type_hints` and  `inspect`?

File: python_modules/dagster/dagster/_core/definitions/source_asset.py [hunk 3]
```diff
@@ -187,6 +178,9 @@ def _get_op_def_compute_fn(self, observe_fn: SourceAssetObserveFunction):
             DecoratedOpFunction,
             is_context_provided,
         )
+        from dagster._core.execution.context.compute import (
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/definitions/source_asset.py:181]: is this file an unrelated change?

File: python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_binding.py [hunk 4]
```diff
@@ -585,3 +585,28 @@ def my_asset(context, my_resource: MyResourceWithDefault):
...
+
+
+class MyModuleLevelResource(ConfigurableResource):
+    str_field: str
+
+
+# Note that an explicit string annotation has the same effect as defining a resource in a module
+# using `from __future__ import annotations`. This test will only work against a module-scoped
+# resource-- this is a hard limitation of string annotations in Python as of 2023-07-06 and Python
+# 3.11.
```
Comment (alangenfeld) [inline python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_binding.py:597]: what does the error look like when you use `from __future__ import annotations` with non module scope type? or i guess just a typo-ed string ?

File: python_modules/dagster/dagster/_core/definitions/inference.py [hunk 5]
```diff
@@ -101,7 +111,8 @@ def _infer_inputs_from_params(
 def infer_input_props(fn: Callable, context_arg_provided: bool) -> Sequence[InferredInputProps]:
     sig = signature(fn)
     params = list(sig.parameters.values())
+    type_hints = get_type_hints(fn)
     descriptions = _infer_input_description_from_docstring(fn)
     params_to_infer = params[1:] if context_arg_provided else params
-    defs = _infer_inputs_from_params(params_to_infer, descriptions=descriptions)
+    defs = _infer_inputs_from_params(params_to_infer, type_hints, descriptions=descriptions)
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/definitions/inference.py:117]: unfortunate how this similar but slightly different information flows through from inspect and typing

**Review by alangenfeld — APPROVED** [review]
seems strictly better than current state, unless `typing_get_type_hints` throws in cases not under test that currently work for users which seems unlikely

File: python_modules/dagster/dagster/_core/definitions/inference.py (hunk 1 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/inference.py:70]: This is the same logic as before but pulling the annotation using `get_type_hints` instead of `inspect.signature`.

File: python_modules/dagster/dagster/_core/definitions/inference.py (hunk 5 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/inference.py:117]: indeed, the python APIs around type annotations are a bit of a mess

File: python_modules/dagster/dagster/_core/definitions/source_asset.py (hunk 3 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/source_asset.py:181]: No. The old code fails with the new method for subtle reasons. When `OpExecutionContext` is defined in a `TYPE_CHECKING` block, it does not exist at runtime. When accessing this annotation through `inspect.signature`, we get a string `"OpExecutionContext"` with no attempt to evaluate because `OpExecutionContext` does not correspond to an input/output/config arg, so it does not fail.

`typing.get_type_hints` attempts to evaluate `OpExecutionContext`, and raises a `NameError`. We avoid this by making it an actual type.

This _is_ technically breaking for users who have done something similar (annotated a context arg with a string referring to a type that is only imported in a `TYPE_CHECKING` block), although it seems unlikely people are doing this since string annotations with ops/asset params will already cause breakage most of the time.

That said I could add some code to catch this error and fall back to old behavior.

File: python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_binding.py (hunk 4 above)
Comment (smackesey) [reply python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_binding.py:597]: I tried to explain the situation in detail here: https://github.com/dagster-io/dagster/pull/14675

But here is an excerpt (you can mentally sub out the quoted annotations `"Foo"` and `"Bar"` with `from __future__ import annotations`-- they have the same effect):

```
from typing import get_type_hints

class Foo:
    ...

def foo(x: "Foo"):
     ...

# works to dereference "Foo" since referent `Foo` is in module-level scope
get_type_hints(foo)  # => {'x': <class '__main__.Foo'>}

def make_bar():
    class Bar:
        ...

    def bar(x: "Bar"):
        ...
    return bar

bar = make_bar()

# does not work to dereference "Bar" since referent `Bar` vanished with local scope
get_type_hints(bar)  # => NameError: name 'Bar' is not defined.
```

File: python_modules/dagster/dagster/_core/definitions/source_asset.py (hunk 3 above)
Comment (alangenfeld) [reply python_modules/dagster/dagster/_core/definitions/source_asset.py:181]: Even if its unlikely its seems like a very terrible error experience so seems worth avoiding. Lets get this case explicitly under test to demonstrate how it errors. 

Should we just try to do the string lookup in globals for the type when we end up with a string from `inspect` instead of eagerly running `get_type_hints`?

File: python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_binding.py (hunk 4 above)
Comment (alangenfeld) [reply python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_binding.py:597]: I was just asking to figure out what the end user experience would be for someone hitting this corner

File: python_modules/dagster/dagster/_core/definitions/source_asset.py (hunk 3 above)
Comment (alangenfeld) [reply python_modules/dagster/dagster/_core/definitions/source_asset.py:181]: Its hard to concoct scenarios where an end user would hit the `OpExecutionContext` problem specifically, but it does seem worth avoiding introducing a known breakage

Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/source_asset.py:181]: >Should we just try to do the string lookup in globals for the type when we end up with a string from inspect instead of eagerly running get_type_hints?

I think `globals()` would be insufficient because it returns the dict for the _current_ module namespace. We need access to the dict for the module namespace where the signature we're processing is written-- `get_type_hints` must access this under the hood somehow.

I think `try`-`except` block that catches `NameError` and falls back to the old behavior if `get_type_hints` fails is the way to go. Note that this will still make code like this (which crashes now) crash:

```
from __future__ import annotations

from dagster import op
from some_lib import SomeType

from typing import TYPE_CHECKING:
     from dagster import OpExecutionContext

@op
def foo(context: OpExecutionContext, bar: SomeType):
    ...
```

This will fail because, when we fall back to the old behavior, `SomeType` will no longer be dereferenced and it will be fed to Dagster machinery as a string.

Comment (alangenfeld) [reply python_modules/dagster/dagster/_core/definitions/source_asset.py:181]: got it - if its possible to throw a better error that would guide users in a helpful direction that would be great

File: python_modules/dagster/dagster/_core/decorator_utils.py (hunk 2 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/decorator_utils.py:55]: yes, `get_type_hints` chokes on `partial`, `inspect.signature` doesn't

File: python_modules/dagster/dagster/_core/definitions/source_asset.py (hunk 3 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/source_asset.py:181]: OK, I pushed a commit that catches `NameError` in `get_type_hints` and gives a thorough error message. Given this error message and the likely extreme rarity of the situation that fallback logic would fix, I don't think it's worth adding that logic.

To spell it out, for the fallback logic to save the user:

1. they would have to  have _only_ the context argument be a string and no other input/output config (which is impossible using `from __future__ import annotations`)
2. have the context type import in a `TYPE_CHECKING` block or local scope

This is very unlikely because the user is almost always going to have other runtime imports from dagster like `op`/`asset`.
